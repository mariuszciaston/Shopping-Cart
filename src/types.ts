export type ProductTypes = {
  id: number;
  image: string;
  title: string;
  price: number;
  description: string;
};

export type ProductsContextType = {
  productsData: ProductTypes[] | null;
  error: null;
  loading: boolean;
};
