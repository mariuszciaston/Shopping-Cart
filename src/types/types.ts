export type Product = {
  id: number;
  image: string;
  title: string;
  price: number;
  description: string;
};

export type ProductState = {
  productsData: Product[] | null;
  error: string | null;
  loading: boolean;
};

export type QuantitySelectorProps = {
  quantity: number;
  setQuantity: (quantity: number) => void;
};

export type CartItem = {
  id: number;
  quantity: number;
};

export type CartState = {
  itemsInCart: CartItem[];
  setItemsInCart: React.Dispatch<React.SetStateAction<CartItem[]>>;
};

export type UpdateCart = (
  id: CartItem["id"],
  quantity: CartItem["quantity"],
  itemsInCart: CartState["itemsInCart"],
  setItemsInCart: CartState["setItemsInCart"],
) => void;

export type ShoppingCart = {
  productsData: Product[] | null;
  itemsInCart: CartItem[];
};
