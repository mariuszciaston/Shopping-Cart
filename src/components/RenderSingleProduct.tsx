import { ProductsContextType } from "@/types";
// import { Link } from "react-router-dom";

const RenderProducts: React.FC<ProductsContextType> = ({
  productsData,
  error,
  loading,
}) => {
  return (
    <>
      {loading && <p>Loading...</p>}
      {error && <p>A network error was encountered</p>}

      <div className="grid grid-cols-[repeat(auto-fill,minmax(340px,1fr))] content-start gap-8">
        {productsData &&
          productsData.map((product) => (
            <div
              key={product.id}
              className="flex flex-col gap-4 rounded-md bg-white p-4 shadow-md"
            >
              <img
                src={product.image}
                alt={product.title}
                className="aspect-square bg-white object-contain"
              />

              <h2 className="font-bold">{product.title}</h2>

              <p>{product.description}</p>

              <p>${product.price}</p>
            </div>
          ))}
      </div>
    </>
  );
};

export default RenderProducts;
