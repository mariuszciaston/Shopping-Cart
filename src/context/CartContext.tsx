import { CartContextType } from "@/types/types";
import { createContext, useContext } from "react";

export const CartContext = createContext<CartContextType | null>(null);

export const useCartContext = () => {
  const context = useContext(CartContext);
  if (!context) {
    throw new Error("useCartContext must be used within a CartProvider");
  }
  return context;
};
