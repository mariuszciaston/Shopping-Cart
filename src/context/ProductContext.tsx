import { ProductsContextType } from "@/types";
import { createContext, useContext } from "react";

export const ProductsContext = createContext<ProductsContextType | null>(null);

export const useProductsContext = () => {
  const context = useContext(ProductsContext);
  if (!context) {
    throw new Error(
      "useProductsContext must be used within a ProductsProvider",
    );
  }
  return context;
};
