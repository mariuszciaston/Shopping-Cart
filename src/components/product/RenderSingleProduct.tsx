import { ProductState } from "@/types/types";
import { Button } from "../ui/Button";
import QuantitySelector from "../ui/QuantitySelector";
import { useState } from "react";
import { useCartContext } from "@/context/CartContext";
import { handleAddToCart } from "../../utils/handleAddToCart";

const RenderSingleProduct: React.FC<ProductState> = ({
  productsData,
  error,
  loading,
}) => {
  const [quantity, setQuantity] = useState(1);
  const { itemsInCart, setItemsInCart } = useCartContext();

  if (loading) return <p>Loading...</p>;
  if (error) return <p>A network error was encountered</p>;

  return (
    <>
      {productsData &&
        productsData.map((product) => (
          <section
            key={product.id}
            className="grid grid-cols-[repeat(auto-fit,minmax(240px,1fr))] gap-8 rounded-md bg-white p-8 shadow-md"
          >
            <img
              src={product.image}
              alt={product.title}
              className="mx-auto aspect-square w-full rounded-md border bg-white object-contain p-8"
            />
            <div className="flex flex-1 flex-col gap-4">
              <h2 className="font-bold">{product.title}</h2>
              <p>{product.description}</p>
              <p>${product.price}</p>

              <QuantitySelector quantity={quantity} setQuantity={setQuantity} />

              <Button
                onClick={() =>
                  handleAddToCart(
                    product.id,
                    quantity,
                    itemsInCart,
                    setItemsInCart,
                  )
                }
                className="mr-auto bg-blue-500 px-8 py-6 text-lg font-bold hover:bg-blue-500"
                variant="default"
              >
                Add to cart
              </Button>
            </div>
          </section>
        ))}
    </>
  );
};

export default RenderSingleProduct;
