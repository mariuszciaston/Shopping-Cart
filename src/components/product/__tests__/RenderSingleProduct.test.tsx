import { fireEvent, render, screen } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import RenderSingleProduct from "../RenderSingleProduct";
import { describe, it, expect, beforeEach, Mock, vi } from "vitest";
import { useCartContext } from "@/context/CartContext";
import { Product } from "@/types/types";

vi.mock("@/context/CartContext", () => ({
  useCartContext: vi.fn(),
}));

const mockUseCartContext = useCartContext as Mock;
const mockSetItemsInCart = vi.fn();

describe("RenderSingleProduct", () => {
  const products: Product[] | null = [
    {
      id: 1,
      title: "Product 1",
      price: 100,
      image: "image1.jpg",
      description: "Description for Product 1",
    },
  ];

  beforeEach(() => {
    mockSetItemsInCart.mockClear();
    mockUseCartContext.mockReturnValue({
      itemsInCart: [],
      setItemsInCart: mockSetItemsInCart,
    });
  });

  it("should display loading message when loading is true", () => {
    render(
      <RenderSingleProduct productsData={[]} loading={true} error={null} />,
    );
    expect(screen.getByText("Loading...")).toBeInTheDocument();
  });

  it("should display error message when error is true", () => {
    render(
      <RenderSingleProduct
        productsData={[]}
        loading={false}
        error={"A network error was encountered"}
      />,
    );
    expect(
      screen.getByText("A network error was encountered"),
    ).toBeInTheDocument();
  });

  it("should display product when data is available", () => {
    expect(products.length).toBeGreaterThan(0);

    render(
      <BrowserRouter>
        <RenderSingleProduct
          productsData={products}
          loading={false}
          error={null}
        />
      </BrowserRouter>,
    );

    products.forEach((product) => {
      const productImg = screen.getByAltText(product.title);
      expect(productImg).toHaveAttribute("src", product.image);
      expect(screen.getByText(product.title)).toBeInTheDocument();
      expect(screen.getByText(product.description)).toBeInTheDocument();
      expect(screen.getByText(`$${product.price}`)).toBeInTheDocument();
    });
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <BrowserRouter>
        <RenderSingleProduct
          productsData={products}
          loading={false}
          error={null}
        />
      </BrowserRouter>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should increase quantity when plus button is clicked", () => {
    render(
      <BrowserRouter>
        <RenderSingleProduct
          productsData={products}
          loading={false}
          error={null}
        />
      </BrowserRouter>,
    );

    const plusButton = screen.getByRole("button", { name: "+" });
    fireEvent.click(plusButton);
    fireEvent.click(plusButton);

    const quantityDisplay = screen.getByRole("spinbutton");
    expect(quantityDisplay).toHaveValue(3);
  });

  it("should decrease quantity when minus button is clicked", () => {
    render(
      <BrowserRouter>
        <RenderSingleProduct
          productsData={products}
          loading={false}
          error={null}
        />
      </BrowserRouter>,
    );

    const plusButton = screen.getByRole("button", { name: "+" });
    fireEvent.click(plusButton);
    fireEvent.click(plusButton);

    const minusButton = screen.getByRole("button", { name: "-" });
    fireEvent.click(minusButton);

    const quantityDisplay = screen.getByRole("spinbutton");
    expect(quantityDisplay).toHaveValue(2);
  });

  it("should not decrease quantity below 1", () => {
    render(
      <BrowserRouter>
        <RenderSingleProduct
          productsData={products}
          loading={false}
          error={null}
        />
      </BrowserRouter>,
    );

    const minusButton = screen.getByRole("button", { name: "-" });
    fireEvent.click(minusButton);
    fireEvent.click(minusButton);
    fireEvent.click(minusButton);

    const quantityDisplay = screen.getByRole("spinbutton");
    expect(quantityDisplay).toHaveValue(1);
  });

  it("should display correct value in quantity input", () => {
    render(
      <BrowserRouter>
        <RenderSingleProduct
          productsData={products}
          loading={false}
          error={null}
        />
      </BrowserRouter>,
    );

    const quantityDisplay = screen.getByRole("spinbutton");
    expect(quantityDisplay).toHaveValue(1);

    const plusButton = screen.getByRole("button", { name: "+" });
    fireEvent.click(plusButton);
    expect(quantityDisplay).toHaveValue(2);

    fireEvent.change(quantityDisplay, { target: { value: "5" } });
    expect(quantityDisplay).toHaveValue(5);
  });

  it("should add item to cart with correct quantity", () => {
    render(
      <BrowserRouter>
        <RenderSingleProduct
          productsData={products}
          loading={false}
          error={null}
        />
      </BrowserRouter>,
    );

    const plusButton = screen.getByRole("button", { name: "+" });
    fireEvent.click(plusButton);
    fireEvent.click(plusButton);

    const addToCartButton = screen.getByRole("button", {
      name: /add to cart/i,
    });
    fireEvent.click(addToCartButton);

    expect(mockSetItemsInCart).toHaveBeenCalledWith([{ id: 1, quantity: 3 }]);
  });

  it("should update quantity if item already exists in cart", () => {
    mockUseCartContext.mockReturnValue({
      itemsInCart: [{ id: 1, quantity: 2 }],
      setItemsInCart: mockSetItemsInCart,
    });

    render(
      <BrowserRouter>
        <RenderSingleProduct
          productsData={products}
          loading={false}
          error={null}
        />
      </BrowserRouter>,
    );

    const addToCartButton = screen.getByRole("button", {
      name: /add to cart/i,
    });
    fireEvent.click(addToCartButton);

    expect(mockSetItemsInCart).toHaveBeenCalledWith([{ id: 1, quantity: 3 }]);
  });
});
