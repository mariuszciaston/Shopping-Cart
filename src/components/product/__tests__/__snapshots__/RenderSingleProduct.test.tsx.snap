// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RenderSingleProduct > matches snapshot 1`] = `
<DocumentFragment>
  <section
    class="grid grid-cols-[repeat(auto-fit,minmax(240px,1fr))] gap-8 rounded-md bg-white p-8 shadow-md"
  >
    <img
      alt="Product 1"
      class="mx-auto aspect-square w-full rounded-md border bg-white object-contain p-8"
      src="image1.jpg"
    />
    <div
      class="flex flex-1 flex-col gap-4"
    >
      <h2
        class="font-bold"
      >
        Product 1
      </h2>
      <p>
        Description for Product 1
      </p>
      <p>
        $100
      </p>
      <div
        class="flex gap-2"
      >
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2 text-lg"
          data-testid="quantity-lower"
        >
          -
        </button>
        <input
          class="flex h-9 rounded-md border border-input bg-transparent px-3 py-1 text-base shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 md:text-sm m-none w-[4rem] pr-0 text-center"
          data-testid="quantity-input"
          max="99"
          min="1"
          type="number"
          value="1"
        />
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2 text-lg"
          data-testid="quantity-boost"
        >
          +
        </button>
      </div>
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 text-primary-foreground shadow h-9 mr-auto bg-blue-500 px-8 py-6 text-lg font-bold hover:bg-blue-500"
      >
        Add to cart
      </button>
    </div>
  </section>
</DocumentFragment>
`;
